api/admin/controller/__init__.py

```diff
@@ -654,6 +663,184 @@ def error_response(self, problem_detail):
+
+    HEAD_TEMPLATE = SignInController.HEAD_TEMPLATE
+
+    RESPONSE_TEMPLATE_WITH_MESSAGE = (
+        response_template_with_message_and_redirect_button.format(
+            head_html=HEAD_TEMPLATE, hr=hr_style, link=small_link_style
+        )
+    )
+
+    def forgot_password(self) -> Union[ProblemDetail, Response]:
+        """Shows forgot password page or starts off forgot password workflow"""
+
+        if not self.admin_auth_providers:
+            return ADMIN_AUTH_NOT_CONFIGURED
+
+        auth = self.admin_auth_provider(PasswordAdminAuthenticationProvider.NAME)
+        if not auth:
+            return ADMIN_AUTH_MECHANISM_NOT_CONFIGURED
+
+        admin = self.authenticated_admin_from_request()
+
+        admin_view_redirect = redirect(self.url_for("admin_view"))
+
+        if isinstance(admin, Admin):
+            return admin_view_redirect
+
+        if flask.request.method == "GET":
+            auth_provider_html = auth.forgot_password_template(admin_view_redirect)
+
+            html = self.FORGOT_PASSWORD_TEMPLATE % dict(
+                auth_provider_html=auth_provider_html,
+                logo_url=AdminClientConfig.lookup_asset_url(key="admin_logo"),
+            )
+            headers = dict()
+            headers["Content-Type"] = "text/html"
+
+            return Response(html, 200, headers)
+
+        admin = self._extract_admin_from_request(flask.request)
+
+        if not admin:
+            return self._response_with_message_and_redirect_button(
+                INVALID_ADMIN_CREDENTIALS.detail,
+                self.url_for("admin_forgot_password"),
+                "Try again",
+                is_error=True,
+                status_code=INVALID_ADMIN_CREDENTIALS.status_code,
+            )
+
+        reset_password_url = self._generate_reset_password_url(admin, auth)
+
+        auth.send_reset_password_email(admin, reset_password_url)
+
+        return self._response_with_message_and_redirect_button(
+            "Email successfully sent! Please check your inbox.",
+            self.url_for("admin_sign_in"),
+            "Sign in",
+        )
+
+    def _extract_admin_from_request(self, request: Request) -> Optional[Admin]:
+        email = request.form.get("email")
+
+        admin = get_one(self._db, Admin, email=email)
+
+        return admin
+
+    def _generate_reset_password_url(
+        self, admin: Admin, auth: PasswordAdminAuthenticationProvider
+    ) -> str:
+        reset_password_token = auth.generate_reset_password_token(admin, self._db)
+
+        reset_password_url = self.url_for(
+            "admin_reset_password",
+            reset_password_token=reset_password_token,
+            _external=True,
+        )
+
+        return reset_password_url
+
+    def reset_password(self, reset_password_token: str) -> Response:
+        """Shows reset password page or process the reset password request"""
+        auth = self.admin_auth_provider(PasswordAdminAuthenticationProvider.NAME)
```

Comment:
Hmm, yeah, that is correct. I'll remove this then, thanks!

Reply:
I went to remove this line and use just the `authenticated_admin_from_request`, however, I need to fetch the PasswordAdminAuthenticationProvider auth provider regardless since it is used on lines 768 and 782. In the end the line count stays the same (although, one call to the admin_auth_provider is saved) but I feel this way seems a bit more explicit.